File: small_problems/medium2/sum_square.rb
# Sum Square - Square Sum
# Write a method that takes an integer n and calculates the difference between the
# square of the sum of the first n positive integers, and the sum of the squares
# of the first n positive integers.

# Input: An Integer, n
# - Represents the endpoint of our square sum/sum square calculations
# - Iterate up to n
# Output: Another Integer
# - Represents the difference (subtraction) between the square of sums and 
#   sum of squares (up to n)

# Square of the sum of the first n (positive) integers:
#   If n = 3; square(1 + 2 + 3) => 36
# Sum of the squares of the first n (positive) integers:
#   If n = 3; sum(1sq + 2sq + 3sq) => 14

# Examples:
# sum_square_difference(3) == 22
#    # -> (1 + 2 + 3)**2 - (1**2 + 2**2 + 3**2)
# sum_square_difference(10) == 2640
# sum_square_difference(1) == 0
# sum_square_difference(100) == 25164150

# Data:
#   - The integer n given as input - acts as endpoint for iteration
#   - Iterate up to n
# Maybe 2 separate variables:
# sum = 0; square_sum = 0
#   Then for each integer up to n, add int to sum and add int squared to sq_sum
#   Square sum at end

# Algorithm:
# Given an integer <n> as input:  (first <n> positive integers ie. start at 1)
# Initialize 2 new variables: sum = 0 and sum_of_squares = 0
# Iterate from 1 up to n. For each integer, int:
#   Increment sum by int.
#   Increment sum_of_squares by int squared
# Calculate the square of sum's value
# Calculate the difference between square of sums and sum of squares

def sum_square_difference(n)
  sum, sum_of_squares = 0, 0
  1.upto(n) do |int|
    sum += int
    sum_of_squares += int ** 2
  end
  square_of_sums = sum ** 2
  square_of_sums - sum_of_squares
end

p sum_square_difference(3) == 22
  # -> (1 + 2 + 3)**2 - (1**2 + 2**2 + 3**2)
p sum_square_difference(10) == 2640
p sum_square_difference(1) == 0
p sum_square_difference(100) == 25164150

File: javaScriptExercises/medium/sumSquare.js
// Sum Square - Square Sum
// Write a function that takes a number (n) and calculates two numbers:
// 1) The square of the sum of the first n positive integers,
// 2) The sum of the squares of the first n positive integers,
// Then returns the difference between 1 and 2 (1 - 2).

// Questions:
// 1) Is it possible for the input n to be missing, negative,
//    or a non-number?
// => Yes => Return undefined
// 2) Is it possible for n to be 0?
// => Yes => Return 0
// 3) Is it possible for Infinity to be an input?
// => Yes => Return Infinity

// Rules / Patterns
// 1) Square of Sums:
// - Sum all integers between 0 and n (0 for 0-inclusivity)
// - Square the resulting sum (**2)
// 2) Sum of Squares:
// - Calculate the square (**2) for each integer between 0..n
// - Sum the resulting squares

// Data: 
// > Input: Number (represents the limit of iteration from 0 - n, 
//   inclusive)
// - validate input:
//   - Must be >= 0
//   - If n = Infinity, return n
// - Create an array of numbers from 0-n
// a) Calculate the sum of the numbers, then square the result
// b) Concurrently transform each number to its square, then sum the
//   results.
// - Calculate and return the difference between a and b.

// > Output: Number (Derived from subtracting Square(sums) - Sum(squares))

// Algorithm:
// Given a number as input, n:

// 1) Validate the input n
// > If n is not a valid integer greater than/equal to 0, return undefined
// > If n is Infinity, return Infinity

// 2) Create an array of all values between 0 to n.

// 3) Calculate the square of sums
// > Sum the numbers in the array created in step 2
// > Square the result

// 4) Calculate the sum of squares
// > Transform the array into an array of squares
// > Then compute the sum from the resulting transformation

// 5) Calculate and return the difference between 3) and 4)

function sumSquareDifference(n) {
  if (!(n >= 0))      return undefined;
  if (n === Infinity) return Infinity;

  let range = Array.from({ length: n + 1 }, (_, i) => i);

  let squareOfSums = sum(range) ** 2;
  let sumOfSquares = sum(range.map(num => num ** 2));

  return squareOfSums - sumOfSquares;
}

function sum(numbers) {
  return numbers.reduce((memo, num) => memo + num, 0);
}

// Refactor: Only use one loop.
// - Initialize two new variables: sum and sumOFSquares
// - Initialize a loop from 1 to n*. For each iteration:
//   * If n = 0 no iteration occurs
// 1) Increment sum by current number
// 2) Increment sumOfSquares by current number ** 2
// 3) After iteration ends, square sum
// 4) Then return squared sum - squareSum

function sumSquareDifference(n) {
  if (!(n >= 0))      return undefined;
  if (n === Infinity) return Infinity;  

  let sum = 0;
  let sumOfSquares = 0;
  for (let i = 1; i <= n; i++) {
    sum += i;
    sumOfSquares += (i ** 2);
  }

  return (sum ** 2) - sumOfSquares;
}

// Examples:
console.log(sumSquareDifference(4));   // 70 (10 ** 2) - (1 + 4 + 9 + 16)
console.log(sumSquareDifference(3));   // 22 --> (1 + 2 + 3)**2 - (1**2 + 2**2 + 3**2)
console.log(sumSquareDifference(10));  // 2640
console.log(sumSquareDifference(1));   // 0
console.log(sumSquareDifference(100)); // 25164150

// Edge Cases:
// 0-Input => 0
console.log(sumSquareDifference(0));   // 0
// Invalid (missing, negative, or NaN) Input => undefined
console.log(sumSquareDifference());         // undefined
console.log(sumSquareDifference(-1));       // undefined
console.log(sumSquareDifference(-Infinity)) // undefined
console.log(sumSquareDifference(NaN));      // undefined
// Infinity => Infinity
console.log(sumSquareDifference(Infinity)); // Infinity
